# HYBRID RECOMMENDER
## BY DWI SMARADAHANA INDRALOKA
***

## Import Library

In [1]:
import pandas as pd

## Upload Dataset

### 1. Movies Detail

In [2]:
data_movie = pd.read_csv("film1.csv")
data_movie.head()

Judul Film     Negara  Rating  Durasi   Genre  Tahun
0  Ada Apa dengan Cinta 2  Indonesia     7.4     123   Drama   2016
1                 Gundala  Indonesia     7.6     123  Action   2019
2              Dilan 1991  Indonesia     6.8     121   Drama   2019
3           Bumi Manusia   Indonesia     7.2     181   Drama   2019
4          Dua Garis Biru  Indonesia     8.2     113   Drama   2019

### 2. People Rating

In [3]:
data_rating = pd.read_csv("film.csv")
data_rating.drop("Timestamp", axis = 1, inplace = True)
data_rating = data_rating.fillna(0)
data_rating.head()

Nama Anda  Ada Apa dengan Cinta 2  Gundala  Dilan 1991  \
0             Hania                     3.0      5.0         4.0   
1  Topik Zulkarnain                     0.0      0.0         0.0   
2    AhokTemanFirli                     0.0      0.0         0.0   
3          franadek                     4.0      4.0         4.0   
4          OM INDRA                     3.0      0.0         2.0   

   Bumi Manusia   Dua Garis Biru  Avengers: End Game  The Lion King  Aladdin  \
0            4.0             4.0                 0.0            0.0      0.0   
1            0.0             0.0                 5.0            5.0      0.0   
2            0.0             0.0                 3.0            0.0      0.0   
3            5.0             3.0                 5.0            4.0      5.0   
4            0.0             5.0                 5.0            0.0      1.0   

   Spiderman: Far From Home  Captain Marvel  
0                       0.0             0.0  
1                       4.0             2.0  
2                       0.0             4.0  
3                       4.0             4.0  
4                       5.0             5.0

## Change Movies Detail's Column "Judul Film" into Index

In [4]:
data_movie.set_index("Judul Film", inplace = True)
data_movie.head()

Negara  Rating  Durasi   Genre  Tahun
Judul Film                                                      
Ada Apa dengan Cinta 2  Indonesia     7.4     123   Drama   2016
Gundala                 Indonesia     7.6     123  Action   2019
Dilan 1991              Indonesia     6.8     121   Drama   2019
Bumi Manusia            Indonesia     7.2     181   Drama   2019
Dua Garis Biru          Indonesia     8.2     113   Drama   2019

## Make Data Dummies for Movies Detail

In [5]:
data_movie1 = pd.get_dummies(data_movie)
data_movie1.head()

Rating  Durasi  Tahun  Negara_Indonesia  Negara_USA  \
Judul Film                                                                    
Ada Apa dengan Cinta 2     7.4     123   2016                 1           0   
Gundala                    7.6     123   2019                 1           0   
Dilan 1991                 6.8     121   2019                 1           0   
Bumi Manusia               7.2     181   2019                 1           0   
Dua Garis Biru             8.2     113   2019                 1           0   

                        Genre_Action  Genre_Adventure  Genre_Animation  \
Judul Film                                                               
Ada Apa dengan Cinta 2             0                0                0   
Gundala                            1                0                0   
Dilan 1991                         0                0                0   
Bumi Manusia                       0                0                0   
Dua Garis Biru                     0                0                0   

                        Genre_Drama  
Judul Film                           
Ada Apa dengan Cinta 2            1  
Gundala                           0  
Dilan 1991                        1  
Bumi Manusia                      1  
Dua Garis Biru                    1

## Normalize the Data Using MinMaxScaler for Movies Detail

In [6]:
from sklearn.preprocessing import MinMaxScaler
data_movie2 = data_movie1.transpose()
col = data_movie1.columns
scaler = MinMaxScaler().fit_transform(data_movie1)
data_movie3 = pd.DataFrame(scaler, columns = col)
data_movie3.head()

Rating    Durasi  Tahun  Negara_Indonesia  Negara_USA  Genre_Action  \
0  0.333333  0.147059    0.0               1.0         0.0           0.0   
1  0.444444  0.147059    1.0               1.0         0.0           1.0   
2  0.000000  0.117647    1.0               1.0         0.0           0.0   
3  0.222222  1.000000    1.0               1.0         0.0           0.0   
4  0.777778  0.000000    1.0               1.0         0.0           0.0   

   Genre_Adventure  Genre_Animation  Genre_Drama  
0              0.0              0.0          1.0  
1              0.0              0.0          0.0  
2              0.0              0.0          1.0  
3              0.0              0.0          1.0  
4              0.0              0.0          1.0

## Calculate Each Movie's Similarity Using Cosine Similarity for Movies Detail

In [7]:
from sklearn.metrics.pairwise import cosine_similarity
data_cosine = cosine_similarity(data_movie3)
data_cosine = pd.DataFrame(data_cosine, columns = data_movie2.columns, index = data_movie2.columns)
data_cosine

Judul Film                Ada Apa dengan Cinta 2   Gundala  Dilan 1991  \
Judul Film                                                               
Ada Apa dengan Cinta 2                  1.000000  0.446439    0.795686   
Gundala                                 0.446439  1.000000    0.647649   
Dilan 1991                              0.795686  0.647649    1.000000   
Bumi Manusia                            0.755808  0.622028    0.892426   
Dua Garis Biru                          0.814795  0.688571    0.910148   
Avengers: End Game                      0.147110  0.645945    0.287912   
The Lion King                           0.026094  0.347188    0.333604   
Aladdin                                 0.034353  0.351610    0.336932   
Spiderman: Far From Home                0.082055  0.693305    0.322753   
Captain Marvel                          0.023062  0.662682    0.336421   

Judul Film                Bumi Manusia   Dua Garis Biru  Avengers: End Game  \
Judul Film                                                                    
Ada Apa dengan Cinta 2         0.755808        0.814795            0.147110   
Gundala                        0.622028        0.688571            0.645945   
Dilan 1991                     0.892426        0.910148            0.287912   
Bumi Manusia                   1.000000        0.830433            0.493865   
Dua Garis Biru                 0.830433        1.000000            0.418739   
Avengers: End Game             0.493865        0.418739            1.000000   
The Lion King                  0.316884        0.341616            0.575243   
Aladdin                        0.356314        0.339206            0.608681   
Spiderman: Far From Home       0.368143        0.411240            0.924322   
Captain Marvel                 0.334299        0.328506            0.836533   

Judul Film                The Lion King   Aladdin  Spiderman: Far From Home  \
Judul Film                                                                    
Ada Apa dengan Cinta 2         0.026094  0.034353                  0.082055   
Gundala                        0.347188  0.351610                  0.693305   
Dilan 1991                     0.333604  0.336932                  0.322753   
Bumi Manusia                   0.316884  0.356314                  0.368143   
Dua Garis Biru                 0.341616  0.339206                  0.411240   
Avengers: End Game             0.575243  0.608681                  0.924322   
The Lion King                  1.000000  0.669125                  0.660515   
Aladdin                        0.669125  1.000000                  0.666612   
Spiderman: Far From Home       0.660515  0.666612                  1.000000   
Captain Marvel                 0.668956  0.671316                  0.969200   

Judul Film                Captain Marvel  
Judul Film                                
Ada Apa dengan Cinta 2          0.023062  
Gundala                         0.662682  
Dilan 1991                      0.336421  
Bumi Manusia                    0.334299  
Dua Garis Biru                  0.328506  
Avengers: End Game              0.836533  
The Lion King                   0.668956  
Aladdin                         0.671316  
Spiderman: Far From Home        0.969200  
Captain Marvel                  1.000000

## Modelling Hybrid Recommender

## 1. Case Person "luck" Want to Watch Movie that Similar with "Captain Marvel"

### 1.1 Show Movies with High Similarity with "Captain Marvel"

In [8]:
data_nama = pd.DataFrame(data_cosine.loc["Captain Marvel"].sort_values(ascending = False))
data_nama = data_nama.iloc[1:6]
data_nama

Captain Marvel
Judul Film                              
Spiderman: Far From Home        0.969200
Avengers: End Game              0.836533
Aladdin                         0.671316
The Lion King                   0.668956
Gundala                         0.662682

### 1.2 Show People Rating for Movies with High Similarity with "Captain Marvel"
* Step 1: Change the data column "Nama Anda" into index
* Step 2: Transpose the data and rename the axis to "Judul Film"
* Step 3: Join high similarity movies data with user rating data using the inner method
* Step 4: Transpose the data and rename the axis to "Nama Anda"

In [9]:
data_rating.set_index("Nama Anda", inplace = True)
data_rating = data_rating.transpose()
data_rating = data_rating.rename_axis(index = "Judul Film")
data_re = pd.merge(data_nama, data_rating, how = "inner", on = "Judul Film")
data_re.drop("Captain Marvel", axis = 1, inplace = True)
data_re = data_re.transpose()
data_re = data_re.rename_axis(index = "Nama Anda")
data_re.head()

Judul Film        Spiderman: Far From Home  Avengers: End Game  Aladdin  \
Nama Anda                                                                 
Hania                                  0.0                 0.0      0.0   
Topik Zulkarnain                       4.0                 5.0      0.0   
AhokTemanFirli                         0.0                 3.0      0.0   
franadek                               4.0                 5.0      5.0   
OM INDRA                               5.0                 5.0      1.0   

Judul Film        The Lion King  Gundala  
Nama Anda                                 
Hania                       0.0      5.0  
Topik Zulkarnain            5.0      0.0  
AhokTemanFirli              0.0      0.0  
franadek                    4.0      4.0  
OM INDRA                    0.0      0.0

### 1.3 Calculate Each Person's Correlation

In [10]:
data_ori = data_re
data_re = data_re.transpose()
data_corr = data_re.corr(method = "pearson")
data_corr

Nama Anda                     Hania  Topik Zulkarnain  AhokTemanFirli  \
Nama Anda                                                               
Hania                      1.000000         -0.604708       -0.250000   
Topik Zulkarnain          -0.604708          1.000000        0.475128   
AhokTemanFirli            -0.250000          0.475128        1.000000   
franadek                  -0.408248         -0.105802        0.612372   
OM INDRA                  -0.475128          0.529851        0.604708   
Mulya                     -0.838525          0.193167        0.279508   
Genjeh                     0.107833         -0.437821        0.377415   
faizah                    -0.250000          0.475128        1.000000   
Dpv                        0.259161          0.216418        0.475128   
Putrisqiana                0.145693          0.310453        0.631336   
Nonton_Saat_Diskon        -0.875000          0.475128        0.375000   
Romantika                       NaN               NaN             NaN   
ANI                       -0.401478         -0.376552       -0.401478   
Rima                       0.250000         -0.475128        0.250000   
bunga                     -1.000000          0.604708        0.250000   
Jawaharal                 -0.612372          0.105802        0.408248   
Febi ganteng gak ada obat -0.153093         -0.238054        0.357217   
Star                       0.051571         -0.142562        0.309426   
Indra Junior              -1.000000          0.604708        0.250000   
Indra 1991 SM              1.000000         -0.604708       -0.250000   
Damar Teman Firli         -0.408248          0.599544        0.612372   
jul                             NaN               NaN             NaN   
luck                      -0.250000          0.475128        1.000000   
Maria O.                  -0.404112         -0.007758       -0.404112   

Nama Anda                  franadek  OM INDRA     Mulya    Genjeh    faizah  \
Nama Anda                                                                     
Hania                     -0.408248 -0.475128 -0.838525  0.107833 -0.250000   
Topik Zulkarnain          -0.105802  0.529851  0.193167 -0.437821  0.475128   
AhokTemanFirli             0.612372  0.604708  0.279508  0.377415  1.000000   
franadek                   1.000000  0.282138  0.684653  0.616316  0.612372   
OM INDRA                   0.282138  1.000000  0.579501  0.484397  0.604708   
Mulya                      0.684653  0.579501  1.000000  0.421962  0.279508   
Genjeh                     0.616316  0.484397  0.421962  1.000000  0.377415   
faizah                     0.612372  0.604708  0.279508  0.377415  1.000000   
Dpv                       -0.105802  0.716418 -0.096583  0.493713  0.475128   
Putrisqiana                0.039653  0.780329  0.000000  0.513209  0.631336   
Nonton_Saat_Diskon         0.612372  0.172774  0.698771 -0.161749  0.375000   
Romantika                       NaN       NaN       NaN       NaN       NaN   
ANI                        0.280976  0.178367  0.705360  0.420555 -0.401478   
Rima                       0.408248  0.475128  0.279508  0.970495  0.250000   
bunga                      0.408248  0.475128  0.838525 -0.107833  0.250000   
Jawaharal                  0.666667  0.775880  0.912871  0.704361  0.408248   
Febi ganteng gak ada obat  0.583333  0.678895  0.627599  0.946485  0.357217   
Star                       0.505291 -0.525697 -0.115316 -0.133465  0.309426   
Indra Junior               0.408248  0.475128  0.838525 -0.107833  0.250000   
Indra 1991 SM             -0.408248 -0.475128 -0.838525  0.107833 -0.250000   
Damar Teman Firli          0.166667  0.987484  0.456435  0.396203  0.612372   
jul                             NaN       NaN       NaN       NaN       NaN   
luck                       0.612372  0.604708  0.279508  0.377415  1.000000   
Maria O.                   0.073324 -0.612864  0.150604 -0.600387 -0.404112   

Nama Anda                       Dpv  Putrisqia

### 1.4 Make Matrix Rating from People with Correlation > 0 with "luck"
* Step 1: Take correlation data with the column "luck" and sorted from largest to smallest
* Step 2: Filter data by only taking correlation values that are more than 0
* Step 3: Join correlation data with user rating data using the inner method

In [11]:
data_rec = pd.DataFrame(data_corr.loc["luck"].sort_values(ascending = False))
data_rec = data_rec[data_rec["luck"] > 0]
data_rec = pd.merge(data_rec, data_ori, how = "inner", on = "Nama Anda")
data_rec

luck  Spiderman: Far From Home  \
Nama Anda                                                       
luck                       1.000000                       0.0   
faizah                     1.000000                       0.0   
AhokTemanFirli             1.000000                       0.0   
Putrisqiana                0.631336                       4.0   
Damar Teman Firli          0.612372                       5.0   
franadek                   0.612372                       4.0   
OM INDRA                   0.604708                       5.0   
Topik Zulkarnain           0.475128                       4.0   
Dpv                        0.475128                       5.0   
Jawaharal                  0.408248                       5.0   
Genjeh                     0.377415                       4.0   
Nonton_Saat_Diskon         0.375000                       4.0   
Febi ganteng gak ada obat  0.357217                       5.0   
Star                       0.309426                       0.0   
Mulya                      0.279508                       4.0   
bunga                      0.250000                       5.0   
Rima                       0.250000                       5.0   
Indra Junior               0.250000                       5.0   

                           Avengers: End Game  Aladdin  The Lion King  Gundala  
Nama Anda                                                                       
luck                                      4.0      0.0            0.0      0.0  
faizah                                    5.0      0.0            0.0      0.0  
AhokTemanFirli                            3.0      0.0            0.0      0.0  
Putrisqiana                               5.0      0.0            0.0      3.0  
Damar Teman Firli                         5.0      0.0            0.0      0.0  
franadek                                  5.0      5.0            4.0      4.0  
OM INDRA                                  5.0      1.0            0.0      0.0  
Topik Zulkarnain                          5.0      0.0            5.0      0.0  
Dpv                                       5.0      0.0            0.0      4.0  
Jawaharal                                 5.0      5.0            4.0      4.0  
Genjeh                                    5.0      5.0            0.0      4.0  
Nonton_Saat_Diskon                        5.0      5.0            5.0      3.0  
Febi ganteng gak ada obat                 5.0      5.0            0.0      3.0  
Star                                      5.0      5.0            5.0      4.0  
Mulya                                     4.0      5.0            2.0      0.0  
bunga                                     5.0      5.0            5.0      4.0  
Rima                                      5.0      5.0            4.0      5.0  
Indra Junior                              5.0      5.0            5.0      0.0

### 1.5 Calculate Total Correlation Value
* Step 1: Transpose the data
* Step 2: Take data with the column "luck"
* Step 3: Do .sum() method to that column

In [12]:
data_rec = data_rec.transpose()
data_corr_film = data_rec.iloc[0,1:]
data_corr_film = pd.DataFrame(data_corr_film)
sum_cor = data_corr_film["luck"].sum()
sum_cor

8.267858969867149

### 1.6 Show the Movie that Hasn't Been Watched by "luck"
* Step 1: Dropping the unused column
* Step 2: Take the movies title that was rated "0" by "luck"
* Step 3: Dropping the unused row

In [13]:
data_rec.drop("luck", axis = 0, inplace = True)
data_rec = data_rec[data_rec["luck"] == 0]
data_rec.drop("luck", axis = 1, inplace = True)
data_rec

Nama Anda                 faizah  AhokTemanFirli  Putrisqiana  \
Spiderman: Far From Home     0.0             0.0          4.0   
Aladdin                      0.0             0.0          0.0   
The Lion King                0.0             0.0          0.0   
Gundala                      0.0             0.0          3.0   

Nama Anda                 Damar Teman Firli  franadek  OM INDRA  \
Spiderman: Far From Home                5.0       4.0       5.0   
Aladdin                                 0.0       5.0       1.0   
The Lion King                           0.0       4.0       0.0   
Gundala                                 0.0       4.0       0.0   

Nama Anda                 Topik Zulkarnain  Dpv  Jawaharal  Genjeh  \
Spiderman: Far From Home               4.0  5.0        5.0     4.0   
Aladdin                                0.0  0.0        5.0     5.0   
The Lion King                          5.0  0.0        4.0     0.0   
Gundala                                0.0  4.0        4.0     4.0   

Nama Anda                 Nonton_Saat_Diskon  Febi ganteng gak ada obat  Star  \
Spiderman: Far From Home                 4.0                        5.0   0.0   
Aladdin                                  5.0                        5.0   5.0   
The Lion King                            5.0                        0.0   5.0   
Gundala                                  3.0                        3.0   4.0   

Nama Anda                 Mulya  bunga  Rima  Indra Junior  
Spiderman: Far From Home    4.0    5.0   5.0           5.0  
Aladdin                     5.0    5.0   5.0           5.0  
The Lion King               2.0    5.0   4.0           5.0  
Gundala                     0.0    4.0   5.0           0.0

### 1.7 Show Recommended Movie with Predicted Rating
* Step 1: Do matrix multiplication to similarity data and people rating for movies that "luck" hasn't watched data
* Step 2: Divide the matrix multiplication results by total similarity value
* Step 3: Show the movies recommendation with the predicted rating value

In [14]:
data_reccomend = data_rec.dot(data_corr_film)
data_reccomend["luck"] = data_reccomend["luck"] / sum_cor
data_reccomend.sort_values("luck", ascending = False)

luck
Spiderman: Far From Home  3.270666
Aladdin                   2.171136
Gundala                   1.822844
The Lion King             1.685959

***
Based on the results above, there are 4 recommended movies for "luck" if he wants to watch a movie that similar to "Captain Marvel" including Spiderman: Far From Home (3.270666), Aladdin (2.171136), Gundala (1.822844) and The Lion King (1.685959).
***

## 2. Case Person "bunga" Want to Watch Movie that Similar with "Dilan 1991"

### 2.1 Show Movies with High Similarity with "Dilan 1991"

In [15]:
data_nama1 = pd.DataFrame(data_cosine.loc["Dilan 1991"].sort_values(ascending = False))
data_nama1 = data_nama1.iloc[1:6]
data_nama1

Dilan 1991
Judul Film                        
Dua Garis Biru            0.910148
Bumi Manusia              0.892426
Ada Apa dengan Cinta 2    0.795686
Gundala                   0.647649
Aladdin                   0.336932

### 2.2 Show People Rating for Movies with High Similarity with "Dilan 1991"
* Step 1: Join high similarity movies data with user rating data using the inner method
* Step 2: Transpose the data and rename the axis to "Nama Anda"

In [16]:
data_re1 = pd.merge(data_nama1, data_rating, how = "inner", on = "Judul Film")
data_re1.drop("Dilan 1991", axis = 1, inplace = True)
data_re1 = data_re1.transpose()
data_re1 = data_re1.rename_axis(index = "Nama Anda")
data_re1.head()

Judul Film        Dua Garis Biru  Bumi Manusia   Ada Apa dengan Cinta 2  \
Nama Anda                                                                 
Hania                        4.0            4.0                     3.0   
Topik Zulkarnain             0.0            0.0                     0.0   
AhokTemanFirli               0.0            0.0                     0.0   
franadek                     3.0            5.0                     4.0   
OM INDRA                     5.0            0.0                     3.0   

Judul Film        Gundala  Aladdin  
Nama Anda                           
Hania                 5.0      0.0  
Topik Zulkarnain      0.0      0.0  
AhokTemanFirli        0.0      0.0  
franadek              4.0      5.0  
OM INDRA              0.0      1.0

### 2.3 Calculate Each Person's Correlation

In [17]:
data_ori = data_re1
data_re1 = data_re1.transpose()
data_corr1 = data_re1.corr(method = "pearson")
data_corr1

Nama Anda                     Hania  Topik Zulkarnain  AhokTemanFirli  \
Nama Anda                                                               
Hania                      1.000000               NaN             NaN   
Topik Zulkarnain                NaN               NaN             NaN   
AhokTemanFirli                  NaN               NaN             NaN   
franadek                  -0.497096               NaN             NaN   
OM INDRA                   0.011990               NaN             NaN   
Mulya                     -0.929981               NaN             NaN   
Genjeh                    -0.492071               NaN             NaN   
faizah                     0.372602               NaN             NaN   
Dpv                        0.323620               NaN             NaN   
Putrisqiana                0.806784               NaN             NaN   
Nonton_Saat_Diskon        -0.643585               NaN             NaN   
Romantika                 -0.058124               NaN             NaN   
ANI                       -0.518079               NaN             NaN   
Rima                      -0.348743               NaN             NaN   
bunga                     -0.459331               NaN             NaN   
Jawaharal                 -0.311741               NaN             NaN   
Febi ganteng gak ada obat -0.643585               NaN             NaN   
Star                      -0.518079               NaN             NaN   
Indra Junior              -0.733946               NaN             NaN   
Indra 1991 SM              0.523114               NaN             NaN   
Damar Teman Firli         -0.058124               NaN             NaN   
jul                       -0.142374               NaN             NaN   
luck                      -0.058124               NaN             NaN   
Maria O.                  -0.929981               NaN             NaN   

Nama Anda                  franadek  OM INDRA     Mulya    Genjeh    faizah  \
Nama Anda                                                                     
Hania                     -0.497096  0.011990 -0.929981 -0.492071  0.372602   
Topik Zulkarnain                NaN       NaN       NaN       NaN       NaN   
AhokTemanFirli                  NaN       NaN       NaN       NaN       NaN   
franadek                   1.000000 -0.799411  0.534522  0.253966 -0.181711   
OM INDRA                  -0.799411  1.000000 -0.206284 -0.231663  0.370668   
Mulya                      0.534522 -0.206284  1.000000  0.475128 -0.582772   
Genjeh                     0.253966 -0.231663  0.475128  1.000000 -0.780329   
faizah                    -0.181711  0.370668 -0.582772 -0.780329  1.000000   
Dpv                       -0.216007 -0.055575 -0.404112  0.612864 -0.287839   
Putrisqiana               -0.327327  0.126323 -0.952579 -0.423207  0.687311   
Nonton_Saat_Diskon         0.337463 -0.230415  0.631336  0.981704 -0.792453   
Romantika                 -0.133631  0.309426 -0.250000  0.475128  0.145693   
ANI                        0.794067 -0.354331  0.324967  0.144375  0.261527   
Rima                       0.534522 -0.593067  0.375000  0.907062 -0.704182   
bunga                      0.384012 -0.518696  0.718421  0.612864 -0.942020   
Jawaharal                 -0.441054  0.382979  0.464140  0.276214 -0.621119   
Febi ganteng gak ada obat  0.337463 -0.230415  0.631336  0.981704 -0.792453   
Star                       0.297775 -0.306448  0.557086  0.986560 -0.865744   
Indra Junior              -0.198517  0.651202  0.557086  0.344895 -0.144291   
Indra 1991 SM             -0.133631 -0.464140 -0.250000  0.259161 -0.582772   
Damar Teman Firli         -0.133631  0.309426 -0.250000  0.475128  0.145693   
jul                       -0.327327  0.126323  0.408248  0.105802 -0.634441   
luck                      -0.133631  0.309426 -0.250000  0.475128  0.145693   
Maria O.                   0.534522 -0.206284  1.000000  0.475128 -0.582772   

Nama Anda                       Dpv  Putrisqia

### 2.4 Make Matrix Rating from People with Correlation > 0 with "bunga"
* Step 1: Take correlation data with the column "bunga" and sorted from largest to smallest
* Step 2: Filter data by only taking correlation values that are more than 0
* Step 3: Join correlation data with user rating data using the inner method

In [18]:
data_rec1 = pd.DataFrame(data_corr1.loc["bunga"].sort_values(ascending = False))
data_rec1 = data_rec1[data_rec1["bunga"] > 0]
data_rec1 = pd.merge(data_rec1, data_ori, how = "inner", on = "Nama Anda")
data_rec1

bunga  Dua Garis Biru  Bumi Manusia   \
Nama Anda                                                            
bunga                      1.000000             0.0            0.0   
Star                       0.733741             0.0            0.0   
Maria O.                   0.718421             0.0            0.0   
Mulya                      0.718421             0.0            0.0   
Febi ganteng gak ada obat  0.671625             0.0            0.0   
Nonton_Saat_Diskon         0.671625             0.0            0.0   
jul                        0.659912             3.0            0.0   
Genjeh                     0.612864             0.0            0.0   
Rima                       0.606168             3.0            4.0   
Jawaharal                  0.565009             5.0            0.0   
Indra 1991 SM              0.493915             0.0            0.0   
franadek                   0.384012             3.0            5.0   
Indra Junior               0.066704             4.0            0.0   

                           Ada Apa dengan Cinta 2  Gundala  Aladdin  
Nama Anda                                                            
bunga                                         0.0      4.0      5.0  
Star                                          4.0      4.0      5.0  
Maria O.                                      0.0      0.0      4.0  
Mulya                                         0.0      0.0      5.0  
Febi ganteng gak ada obat                     4.0      3.0      5.0  
Nonton_Saat_Diskon                            4.0      3.0      5.0  
jul                                           0.0      3.0      3.0  
Genjeh                                        5.0      4.0      5.0  
Rima                                          5.0      5.0      5.0  
Jawaharal                                     2.0      4.0      5.0  
Indra 1991 SM                                 0.0      5.0      0.0  
franadek                                      4.0      4.0      5.0  
Indra Junior                                  4.0      0.0      5.0

### 2.5 Calculate Total Correlation Value
* Step 1: Transpose the data
* Step 2: Take data with the column "bunga"
* Step 3: Do .sum() method to that column

In [19]:
data_rec1 = data_rec1.transpose()
data_corr_film1 = data_rec1.iloc[0,1:]
data_corr_film1 = pd.DataFrame(data_corr_film1)
sum_cor1 = data_corr_film1["bunga"].sum()
sum_cor1

6.902416550630752

### 2.6 Show the Movie that Hasn't Been Watched by "bunga"
* Step 1: Dropping the unused column
* Step 2: Take the movies title that was rated "0" by "bunga"
* Step 3: Dropping the unused row

In [20]:
data_rec1.drop("bunga", axis = 0, inplace = True)
data_rec1 = data_rec1[data_rec1["bunga"] == 0]
data_rec1.drop("bunga", axis = 1, inplace = True)
data_rec1

Nama Anda               Star  Maria O.  Mulya  Febi ganteng gak ada obat  \
Dua Garis Biru           0.0       0.0    0.0                        0.0   
Bumi Manusia             0.0       0.0    0.0                        0.0   
Ada Apa dengan Cinta 2   4.0       0.0    0.0                        4.0   

Nama Anda               Nonton_Saat_Diskon  jul  Genjeh  Rima  Jawaharal  \
Dua Garis Biru                         0.0  3.0     0.0   3.0        5.0   
Bumi Manusia                           0.0  0.0     0.0   4.0        0.0   
Ada Apa dengan Cinta 2                 4.0  0.0     5.0   5.0        2.0   

Nama Anda               Indra 1991 SM  franadek  Indra Junior  
Dua Garis Biru                    0.0       3.0           4.0  
Bumi Manusia                      0.0       5.0           0.0  
Ada Apa dengan Cinta 2            0.0       4.0           4.0

### 2.7 Show Recommended Movie with Predicted Rating
* Step 1: Do matrix multiplication to similarity data and people rating for movies that "bunga" hasn't watched data
* Step 2: Divide the matrix multiplication results by total similarity value
* Step 3: Show the movies recommendation with the predicted rating value

In [21]:
data_reccomend1 = data_rec1.dot(data_corr_film1)
data_reccomend1["bunga"] = data_reccomend1["bunga"] / sum_cor1
data_reccomend1.sort_values("bunga", ascending = False)

bunga
Ada Apa dengan Cinta 2  2.511585
Dua Garis Biru          1.165119
Bumi Manusia            0.629451

***
Based on the results above, there are 3 recommended movies for "bunga" if she wants to watch a movie that similar to "Dilan 1991" including Ada Apa dengan Cinta 2 (2.511585), Dua Garis Biru (1.165119) dan Bumi Manusia (0.629451).
***

## Conclusion
By using Hybrid Recommender we can use two data as inputs, the inputs are the name of the person and the title of the movie that they are interested in. From that, we can give the movies recommendation that can be watched based on the similarity between the movie with the other movies and based on the correlation between the person with the other people. In the real case, we can apply this algorithm to display the movies recommendation on a movie website when someone has logged in and has searched for a movie on the website.

***
# THANK YOU